src/messaging/ExchangeContext.cpp

```diff
@@ -92,6 +93,32 @@ CHIP_ERROR ExchangeContext::SendMessage(uint16_t protocolId, uint8_t msgType, Pa
     // Set the message type for this header.
     payloadHeader.SetMessageType(msgType);
 
+    payloadHeader.SetInitiator(IsInitiator());
+
+    // If sending via UDP and the auto-request ACK feature is enabled, automatically
+    // request an acknowledgment, UNLESS the NoAutoRequestAck send flag has been specified.
+    if (mReliableMessageContext.AutoRequestAck() && !sendFlags.Has(SendMessageFlags::kSendFlag_NoAutoRequestAck))
+    {
+        payloadHeader.SetNeedsAck(true);
+    }
+
+    // Abort early if Throttle is already set.
+    VerifyOrExit(!mReliableMessageContext.IsThrottling(), err = CHIP_ERROR_SEND_THROTTLED);
+
+    // If there is a pending acknowledgment piggyback it on this message.
+    if (mReliableMessageContext.HasPeerRequestedAck())
+    {
+        payloadHeader.SetAckId(mReliableMessageContext.mPendingPeerAckId);
+
+        // Set AckPending flag to false after setting the Ack flag;
+        mReliableMessageContext.SetAckPending(false);
+
+#if !defined(NDEBUG)
+        ChipLogProgress(ExchangeManager, "Piggybacking Ack for MsgId:%08" PRIX32 " with msg",
+                        mReliableMessageContext.mPendingPeerAckId);
+#endif
+    }
+
```

Comment:
Would you please extract these logics into a function in `ReliableMessageContext` or `ReliableMessageManager` class ?

The code is totally fine, and reasonable. But putting the code here will significantly distract one's mind when reading the code.

Reply:
We need to pass payloadHeader in and out of ReliableMessageContext, currently, ReliableMessageContext/ReliableMessageManager only contains the CRMP processing logic. This separation looks good to me,  I have concern if we allow CRMP modules get involved in packet header processing since payloadHeader is not prepared in a central place in this way. 

I agree we have some room to further optimize, but I prefer revisit this item after the first integration since there are a lot of unit functions can be combined if we allow CRMP to handle the packet processing.